tests/components/tradfri/test_cover.py

```diff
@@ -0,0 +1,219 @@
+    assert cover_1 is not None
+    assert cover_1.state == "open"
+    assert cover_1.attributes["current_position"] == 60
+
+
+async def test_cover_observed(hass, mock_gateway, mock_api_factory):
+    """Test that covers are correctly observed."""
+    state = {
+        "current_cover_position": 1,
+    }
+
+    cover = mock_cover(test_state=state)
+    mock_gateway.mock_devices.append(cover)
+    await setup_integration(hass)
+    assert len(cover.observe.mock_calls) > 0
+
+
+async def test_cover_available(hass, mock_gateway, mock_api_factory):
+    """Test cover available property."""
+
+    cover = mock_cover(test_state={"current_cover_position": 1}, device_number=1)
+    cover.reachable = True
+
+    cover2 = mock_cover(test_state={"current_cover_position": 1}, device_number=2)
+    cover2.reachable = False
+
+    mock_gateway.mock_devices.append(cover)
+    mock_gateway.mock_devices.append(cover2)
+    await setup_integration(hass)
+
+    assert hass.states.get("cover.tradfri_cover_1").state == "open"
+    assert hass.states.get("cover.tradfri_cover_2").state == "unavailable"
+
+
+def create_all_set_position_on_cases():
+    """Create all test cases."""
+    all_test_cases = [
+        ["test_features", "test_data", "expected_result", "device_id"],
+        [],
+    ]
+    index = 1
+    for test_case in SET_POSITION_TEST_CASES:
+        case = deepcopy(test_case)
+        case.append(index)
+        index += 1
+        all_test_cases[1].append(case)
+
+    return all_test_cases
+
+
+@pytest.mark.parametrize(*create_all_set_position_on_cases())
+async def test_set_cover_position(
+    hass,
+    mock_gateway,
+    mock_api_factory,
+    test_features,
+    test_data,
+    expected_result,
+    device_id,
+):
+    """Test setting position a cover."""
+    # Note pytradfri style, not hass. Values not really important.
+    initial_state = {
+        "current_cover_position": 0,
+    }
+
+    # Setup the gateway with a mock cover.
+    cover = mock_cover(
+        test_features=test_features, test_state=initial_state, device_number=device_id
+    )
+    mock_gateway.mock_devices.append(cover)
+    await setup_integration(hass)
+
+    # Use the turn_on service call to change the cover state.
+    await hass.services.async_call(
+        "cover",
+        "set_cover_position",
+        {"entity_id": f"cover.tradfri_cover_{device_id}", **test_data},
+        blocking=True,
+    )
+    await hass.async_block_till_done()
+
+    # Check that the cover is observed.
+    mock_func = cover.observe
+    assert len(mock_func.mock_calls) > 0
+    _, callkwargs = mock_func.call_args
+    assert "callback" in callkwargs
+    # Callback function to refresh cover state.
+    callback = callkwargs["callback"]
+
+    responses = mock_gateway.mock_responses
+    # State on command data.
+    data = {"15015": [{"5536": 50.0}]}
+    # Add data for all sent commands.
+    for resp in responses:
+        data["15015"][0] = {**data["15015"][0], **resp["15015"][0]}
```

Comment:
Looks like this just inserts data in the same position on each loop iteration.

Reply:
In short, what I believe it does is emulate a response from the gateway before sending the device to the `callback` function further down. The test fails if I remove the loop.

What we are doing is setting correct properties for the `[raw](https://github.com/home-assistant-libs/pytradfri/blob/ef2614b0ccdf628abc3b9559dc00b95ec6e4bd72/pytradfri/device/blind.py#L23
)` property in the `Blind` class that is handled a few lines down.

Reply:
How many items is there in `responses`? If we only need to set the last item in data, we can just access the last item directly without a loop.